### PR TITLE
Adds an asynchronous queue that sends events

### PR DESCRIPTION
**Pull request by RomainBoulay**
### Short description 📝

> Foundation for analytics.

### Solution 📦

> Adds an asynchronous queue that sends events to diverse providers.

### Implementation 👩‍💻👨‍💻

- [x] Adds an asynchronous queue
- [x] Provides the ability to store events that could not have been sent on disk, in order to try to send them again at a later point
- [x] Enables contributors to add diverse dispatchers (i.e. analytics providers)  
- [x] Add tests

### CODE REVIEW

File: Sources/TuistAsyncQueue/AsyncQueue.swift [hunk 1]
```diff
@@ -0,0 +1,160 @@
+import Foundation
+import Queuer
+import RxSwift
```
Comment (pepibumur) [inline Sources/TuistAsyncQueue/AsyncQueue.swift:3]: I don't think we need this anymore, do we?

File: Sources/TuistAsyncQueue/AsyncQueuePersistor.swift [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                try FileHandler.shared.delete(path)
+                observer(.completed)
+            } catch {
+                observer(.error(error))
+            }
+            return Disposables.create()
+        }
+    }
+
+    func readAll() -> Single<[AsyncQueueEventTuple]> {
```
Comment (pepibumur) [inline Sources/TuistAsyncQueue/AsyncQueuePersistor.swift:70]: @RomainBoulay what do you think about adding a bit of documentation somewhere in this file where we describe the convention that we follow for persisting the events. It'll make it easier for developers in the future to understand the logic in this method.

**Review by pepibumur — APPROVED** [review]

File: Sources/TuistCore/AsyncQueue/AsyncQueueDispatcher.swift [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import Foundation
+import RxSwift
+
+/// Async queue dispatcher.
+public protocol AsyncQueueDispatcher {
```
Comment (pepibumur) [inline Sources/TuistCore/AsyncQueue/AsyncQueueDispatcher.swift:5]: @RomainBoulay assuming the `dispatchPersisted` will de-serialize the data into an instance of `AsyncQueueEvent` using the type associated with this dispatcher. Do you think we could define a protocol extension on `AsyncQueueDispatcher` to provide a default implementation? Or do you think we'd end up with the same issue that we ran into with the usage of generics and protocols that have an associated type?

File: Sources/TuistAsyncQueue/AsyncQueue.swift (hunk 1 above)
Comment (RomainBoulay) [reply Sources/TuistAsyncQueue/AsyncQueue.swift:3]: RxSwift is still needed by this class

Comment (RomainBoulay) [reply Sources/TuistAsyncQueue/AsyncQueue.swift:3]: (see DisposeBag)

File: Sources/tuist/main.swift [hunk 4]
```diff
@@ -8,4 +9,7 @@ if CommandLine.arguments.contains("--verbose") { try? ProcessEnv.setVar(Constant
 LogOutput.bootstrap()
 
 import TuistKit
-TuistCommand.main()
+
+try AsyncQueue.run {
+    TuistCommand.main()
+}
```
Comment (RomainBoulay) [inline Sources/tuist/main.swift:15]: @pepibumur I have to remove this as it fails some acceptance tests (I'm looking at the scaffolding ones).
Could you explain me what you had in mind with this code here? Thanks!

Comment (RomainBoulay) [reply Sources/tuist/main.swift]: Link to failed AT: https://github.com/tuist/tuist/pull/1930/checks?check_run_id=1336080654

**Pull request merged**